docs/website/docs/v1.x/dsl/middleware.md

```diff
@@ -3,4 +3,422 @@ sidebar_position: "8"
 ---
 # Middleware
 
-WIP
\ No newline at end of file
+## What is a "Middleware" and why they are needed
+Before introducing middleware, let us understand why they are needed.
+
+Consider the following example where we have two endpoints within HttpApp
+* GET user by id and
+* GET multiple users paginated
+```scala
+  private val app = Http.collectZIO[Request] {
+    case Method.GET -> !! / "users" / id =>
+      // core business logic  
+      dbService.lookupUsersById(id).map(Response.json(_.json))
+    case Method.GET -> !! / "users"    =>
+      // core business logic  
+      dbService.paginatedUsers(pageNum).map(Response.json(_.json))
+  }
+```
+#### The polluted code violates the principle of "Separation of concerns"
+
+As our application grows, we want to code the following aspects like
+* Basic Auth
+* Request logging
+* Response logging
+* Timeout and retry
+
+For both our example endpoints, our core business logic gets buried under boilerplate like this
+
+```scala
+            (for {
+                // validate user
+                _    <- MyAuthService.doAuth(request)
+                // log request
+                _    <- logRequest(request)
+                // core business logic
+                user <- dbService.lookupUsersById(id).map(Response.json(_.json))
+                resp <- Response.json(user.toJson)
+                // log response
+                _    <- logResponse(resp)                
+            } yield resp)
+                    .timeout(2.seconds)
+                    .retryN(5)
+```
+Imagine repeating this for all our end points!!!
+
+So there are two problems with this approach
+* We are dangerously coupling our business logic with a lower level concern (like applying timeouts)
+* Also, we will have to do it for every single route in the system. For 100 routes we will need to repeat 100 timeouts!!!
+
+This can lead to a lot of boilerplate clogging our neatly written endpoints affecting readability, thereby leading to maintenance costs.
```

Comment:
```suggestion
This can lead to a lot of boilerplate clogging our neatly written endpoints affecting readability, thereby leading to increased maintenance costs.
```